clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -264,40 +271,133 @@ public boolean rename(Path src, Path dst) throws IOException {
+            srcStatus = getFileStatus(src);
+        } catch (FileNotFoundException e) {
+            LOG.error("rename: src {} does not exist, rename failed.", src, e);
             return false;
         }
+        if (!srcStatus.isDirectory()) {
+            return renameFile(srcStatus, dst);
+        }
+        return renameDirectory(src, dst);
+    }
 
-        return renameObject(srcStat, srcObjectLoc, dstObjectLoc);
+    /**
+     * Recursively rename objects under src dir.
+     * @return true if all objects under src renamed successfully, false otherwise.
+     */
+    private boolean renameDirectory(Path src, Path dst) throws IOException {
+        boolean dstExists = false;
+        LakeFSFileStatus dstFileStatus;
+        try {
+            // May be unnecessary with https://github.com/treeverse/lakeFS/issues/1691
+            dstFileStatus = getFileStatus(dst);
+            dstExists = true;
+            if (!dstFileStatus.isDirectory()) {
+                LOG.error("renameDirectory: dst {} is not a directory, rename of directory to a file " +
+                        "is unsupported", dst);
+                return false;
+            }
+        } catch (FileNotFoundException e) {
+            LOG.debug("renameDirectory: dst {} does not exist", dst);
+        }
+
+        ListingIterator iterator = new ListingIterator(src, true, listAmount);
+        while (iterator.hasNext()) {
+            LakeFSLocatedFileStatus locatedFileStatus = iterator.next();
+            Path objDst = dstExists ? buildObjPathOnExistingDestinationDir(locatedFileStatus.getPath(), dst) :
+                    buildObjPathOnNonExistingDestinationDir(locatedFileStatus.getPath(), src, dst);
+            try {
+                if (!renameObject(locatedFileStatus.toLakeFSFileStatus(), objDst)) {
+                    throw new IOException();
+                }
+            } catch (IOException e) {
+                // Rename dir operation in non-transactional. if one object rename failed we will end up in an
+                // intermediate state.
+                LOG.error("renameDirectory: failed to rename src dir {}", src);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * TODO (Tals): move this description into a component-test
+     * Sample input and output
+     * input:
+     * renamedObj: lakefs://repo/main/dir1/file1.txt
+     * srcDirPath: lakefs://repo/main/dir1
+     * dstDirPath: lakefs://repo/main/dir2
+     * output:
+     * lakefs://repo/main/dir2/file1.txt
+     */
+    private Path buildObjPathOnNonExistingDestinationDir(Path renamedObj, Path srcDir, Path dstDir) {
+        ObjectLocation renamedObjLoc = pathToObjectLocation(renamedObj);
+        ObjectLocation srcDirLoc = pathToObjectLocation(srcDir);
+        ObjectLocation dstDirLoc = pathToObjectLocation(dstDir);
+        String formattedPath = renamedObjLoc.getPath().replace(srcDirLoc.getPath(), dstDirLoc.getPath());
```

Comment:
This (and any unanchored string search+replace) is unsafe when paths repeat themselves.